src/reflect/scala/reflect/internal/Types.scala

```diff
@@ -2408,7 +2408,7 @@ trait Types
     override final def equals(other: Any): Boolean = {
       if (this eq other.asInstanceOf[AnyRef]) true
       else other match {
-        case otherTypeRef: TypeRef =>
+        case otherTypeRef: TypeRef if !(otherTypeRef.isInstanceOf[ObjectTpeJavaRef] || this.isInstanceOf[ObjectTpeJavaRef]) =>
```

Comment:
can we use `eq`/`ne` here instead of the type test?

Reply:
I tried that, but it caused cycles.